### ISSUE TITLE
XOR highlight is not practical when working on low color images

### ISSUE THREAD

**Issue by GoogleCodeExporter** [birdal46/grafx2#185]
```
When you use a small number of colors (ie 16), the XOR lines we use to 
highlight selection (when drawing a rectangle or capturing a brush) often 
gives color indexes that end up showing all black. This mean the selection 
box is not visible.
We can tell people to put an ugly color (like #00FF00) in the unused 
palette indexes, but this is not really a clean solution.
We can make the program look for the best opposite color (ie find nearest 
match to #FF00FF as opposite to #00FF00) but that's an heay process.
We can make the selection box use dashed CM_WHITE/CM_BLACK lines, which 
seems a good compromise (should be visible EXCEPT if you use exactly these 
colors in a pattern), fast enough to draw, and it seems to be what most 
other program are using (een our "brush grab" icon suggest it).

Any ideas on this ?
```

Original issue reported on code.google.com by `pulkoma...@gmail.com` on 3 Jul 2009 at 10:28

**Comment by GoogleCodeExporter** [comment]
```
As a user, when I do a new 8- or 16-color piece with a brand new palette that I
started "all-black", if I have the problem I just set a white color on the 
palette
entry that is opposite to my own (used) black. It's an unused color so not a 
problem.
I may have to repeat the process if I rearrange my palette, but after that, it's
solved and remembered for the image.
I just tested: when you have a finalized 16-color palette, it's easy to copy 
them as
last 16 colors, invert range, negative it, and then the XOR lines produce 
exactly the
negative of the RGB values :)

Note that after loading a native low-colors image, pressing "backspace" in 
Palette
screen will often solve the issue as it puts white in #FF, and two greys in #FE 
and #FD.

Dotted black-white has the problem that they completely hide what they 
cover...(in
zoomed view for example, it will be very fat lines.) The XOR, on the other hand,
still shows the shape with different colors; it's very useful when you want to 
cut a
sprite to its minimal bounding rectangle, it helps you avoid picking an extra 
empty
line or column.

```

Original comment by `yrizoud` on 4 Jul 2009 at 12:38

**Comment by GoogleCodeExporter** [comment]
```
One thing we can do 'for free' is to XOR on something else than 0xFF. For 
example,
with a 16-color image (in 16 first palette entries), XORing on 0x0F will revert 
the
colors : 0 <-> 15 , 1 <-> 14 etc.
```

Original comment by `yrizoud` on 4 Jul 2009 at 7:18

**Comment by GoogleCodeExporter** [comment]
```
That would work fine. Ideally we would have to find the value to use by looking 
for 
biggest used colorindex in picture, or maybe biggest non-black color id. This 
second 
solution will not work so well as I tend to have grafx2 menu's colors at the 
end of 
the palette. But the first one seems a bit heavy to find, and more importantly, 
when 
do we update it ? each time a pixel is written is not reasonable, but finding 
it 
only once means you will always find only the bg color is used.

Alternatively, I was thinking of adding 8 to the color index instead of doing 
an 
XOR. This way at least half of the 16color palette get a visible equivalent. 
Using a 
value less than 8 means you get more visible colors, but more odds of replacing 
them 
with a color in a shade that's pretty near, so not really more visible.

Finally, we could also build a mapping of "most different color" calculated 
from RGB 
indexes and use that to find the color we use for XOR. Then the same table 
could be 
used by some new drawing effect...
```

Original comment by `pulkoma...@gmail.com` on 4 Jul 2009 at 7:29

**Comment by GoogleCodeExporter** [comment]
```
What about this : We define 3 or 4 methods (XOR 255, XOR 15, ADD 8...). Each 
time you
exit the palette menu, the program checks all of them and picks the best.
To check a method, it proceeds by:
- For each USED color in image:
  - compute the resulting color after transform,
  - get its rgb value,
  - compute the heuristic based on how 'different' the RGB is. Identical = very bad, etc.
```

Original comment by `yrizoud` on 4 Jul 2009 at 8:00

**Comment by GoogleCodeExporter** [comment]
```
Mh... as there are only 256 colors i think it's faster to calculate the 256 
"most 
different colors" and create a lookup table than test all the methods. However, 
using the lookup table will be slightly slower than a simple xor operation.
Calculate 4 or 5 different XOR methods, then computing the differences in RGB, 
then 
finding the average difference and then find the method with the bigger score 
sounds 
like an heavy task. I think finding the most different color for each of the 
colors 
can be done easily in an Nlog(N) complexity algorithm or perhaps a little less 
: you 
have to comapre 255 colors with the first one, for the second one it is either 
already the most different from 1 and you're done, or you have 254 colors to 
compare 
it to. Overall, i think in most cases, it's quite fast.
```

Original comment by `pulkoma...@gmail.com` on 4 Jul 2009 at 8:16

**Comment by GoogleCodeExporter** [comment]
```
As you wish, but you'll have to find an algorithm. And since the lookup method 
is not
reversible, several operations which used XOR to undo their changes will need 
some
more changes to be able to use it.

```

Original comment by `yrizoud` on 5 Jul 2009 at 6:31

**Comment by GoogleCodeExporter** [comment]
```
what about :

colormap[256]

for(i=0;i<256;i++)colormap[i]=i;

for( i=0 ; i <256; i++)
{
  if colormap[i]!=i continue; //color already initialized
  acd=0;
  for( j=i;j<256;j++)
  {
    k=AbsoluteColorDifference(palette[i],palette[j]);
    if(k>acd)
    {
      acd=k; // new best difference
      colormap[i]=j; // new best opposite color index
      colormap[j]=i;
    }
  }
}

(minus potential infinite loop palette when there is only one color in the 
palette... who would do that anyway ?)

Then to draw an "xor" line you just replace index i by index colormap[i]. he 
table 
is reversible so applying the change again undo it. I don't think we need to 
change 
any operation. The cursor is erased when opening a window, right ? we only need 
to 
generate this able on file loading and opening the palette screen.
```

Original comment by `pulkoma...@gmail.com` on 5 Jul 2009 at 6:40

**Comment by GoogleCodeExporter** [comment]
```
The problem is that the first colors will pick their best opposite, while the 
last 
colors will have to do with what's remaining. If the colors are somewhat sorted 
(and 
they often are), you'll get black <-> white, then dark <-> light, then ... 
middle <-
> middle.

I was thinking of picking the method with the best "worst case".

```

Original comment by `yrizoud` on 6 Jul 2009 at 11:24

**Comment by GoogleCodeExporter** [comment]
```
Mh, yes, I didn't think of that... We will have to go for the try-and-score 
multiple 
methods, then, as I can't think of anything simple.
```

Original comment by `pulkoma...@gmail.com` on 6 Jul 2009 at 11:30

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
Original comment by `pulkoma...@gmail.com` on 15 Sep 2009 at 7:13
- Added labels: **Milestone-2.3**

**Comment by GoogleCodeExporter** [comment]
```
Issue 287 has been merged into this issue.
```

Original comment by `pulkoma...@gmail.com` on 17 Jan 2010 at 9:28

**Comment by GoogleCodeExporter** [comment]
```
We already have a setting for the grid XOR value. Would it be a maked 
improvement if I made the XOR cursor align itself on this same parameter 
(instead of the fixed 255 at present) ?
The change would be for hairlines, brush-grabbing, lines and other geometric 
shapes, contour fill, brush transformation handles and outlines...
```

Original comment by `yrizoud` on 2 Sep 2010 at 11:28

**Comment by GoogleCodeExporter** [comment]
Original comment by `pulkoma...@gmail.com` on 15 Feb 2011 at 8:12
- Added labels: **Milestone-2.4**

**Comment by GoogleCodeExporter** [comment]
```
Arbitering issues that make it to v2.4
```

Original comment by `yrizoud` on 8 Mar 2012 at 7:18
- Added labels: **Milestone-2.5**

**Comment by GoogleCodeExporter** [comment]
```
I think both the XOR grid and the XOR highlight should use the same values, 
and, if possible, have an "Automatic" setting that would adapt to how many 
colors the image actually has. (that is, if it has all 256 colors, set 255; if 
64, set 63, and so on. This is not a perfect solution, but a working one that 
wouldn't be too hard to implement.)
```

Original comment by `m.zub...@gmail.com` on 30 Dec 2012 at 7:46

**Comment by GoogleCodeExporter** [comment]
```
Fixed in r2048.
```

Original comment by `pulkoma...@gmail.com` on 6 Jan 2013 at 12:47
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
r2048 doesn't resolve the problem fully. Both XOR grid and XOR highlight for 
this image seem to be completely black.

(Also, XOR grid also needs to be fixed.)
```

Original comment by `m.zub...@gmail.com` on 2 Feb 2013 at 7:12

Attachments:
- [valgrind3_test.gif](https://storage.googleapis.com/google-code-attachments/grafx2/issue-185/comment-18/valgrind3_test.gif)

**Comment by GoogleCodeExporter** [comment]
```
Ok, you can try r2049.
```

Original comment by `pulkoma...@gmail.com` on 2 Feb 2013 at 8:38

**Comment by GoogleCodeExporter** [comment]
```
Seems like r2049 has opened up a new problem. Now, the XOR highlight seems to 
leave a trail of a different color. For example, on the image attached here, 
the blue pixels turn brown.
```

Original comment by `m.zub...@gmail.com` on 3 Feb 2013 at 9:12

Attachments:
- [highdigits.gif](https://storage.googleapis.com/google-code-attachments/grafx2/issue-185/comment-20/highdigits.gif)

**Comment by GoogleCodeExporter** [comment]
```
The problem seems to be that, as a result, some colors tend to have the same 
xor_lut values. That, in itself, is completely fine.

But Grafx2's cursor code seems to xor the pixels once again instead of 
redrawing them, and these pixels get redrawn into wrong colors as a result.
```

Original comment by `m.zub...@gmail.com` on 3 Feb 2013 at 12:41

**Comment by GoogleCodeExporter** [comment]
```
Colors should never have the same XOR value, precisely because we XOR again to 
erase (this is fastest, we don't have to save the original color).
Anyway, r2051 should be better.
```

Original comment by `pulkoma...@gmail.com` on 3 Feb 2013 at 6:16
